File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Test/SFSDKTestRequestListener.m
/*
 Copyright (c) 2012-present, salesforce.com, inc. All rights reserved.
 
 Redistribution and use of this software in source and binary forms, with or without modification,
 are permitted provided that the following conditions are met:
 * Redistributions of source code must retain the above copyright notice, this list of conditions
 and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice, this list of
 conditions and the following disclaimer in the documentation and/or other materials provided
 with the distribution.
 * Neither the name of salesforce.com, inc. nor the names of its contributors may be used to
 endorse or promote products derived from this software without specific prior written
 permission of salesforce.com, inc.
 
 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#import "SFSDKTestRequestListener.h"
#import "TestSetupUtils.h"
#import "SFUserAccountManager.h"

NSString* const kTestRequestStatusWaiting = @"waiting";
NSString* const kTestRequestStatusDidLoad = @"didLoad";
NSString* const kTestRequestStatusDidFail = @"didFail";

@interface SFSDKTestRequestListener ()
@end

@implementation SFSDKTestRequestListener

@synthesize dataResponse = _dataResponse;
@synthesize lastError = _lastError;
@synthesize returnStatus = _returnStatus;
@synthesize maxWaitTime = _maxWaitTime;

- (id)init
{
    self = [super init];
    if (nil != self) {
        self.maxWaitTime = 30.0;
        self.returnStatus = kTestRequestStatusWaiting;
    }
    return self;
}

- (void)dealloc {
    self.dataResponse = nil;
    self.lastError = nil;
    self.returnStatus = nil;
}

- (NSString *)waitForCompletion {
    NSDate *startTime = [NSDate date] ;
    while ([self.returnStatus isEqualToString:kTestRequestStatusWaiting]) {
        NSTimeInterval elapsed = [[NSDate date] timeIntervalSinceDate:startTime];
        if (elapsed > self.maxWaitTime) {
            [SFSDKCoreLogger d:[self class] format:@"Request took too long (> %f secs) to complete.", elapsed];
            return kTestRequestStatusDidFail;
        }
        [[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
    }
    return self.returnStatus;
}

#pragma mark - SFIdentityCoordinatorDelegate

- (void)identityCoordinatorRetrievedData:(SFIdentityCoordinator *)coordinator
{
    [SFSDKCoreLogger i:[self class] format:@"%@", NSStringFromSelector(_cmd)];
    self.returnStatus = kTestRequestStatusDidLoad;
}

- (void)identityCoordinator:(SFIdentityCoordinator *)coordinator didFailWithError:(NSError *)error
{
    [SFSDKCoreLogger i:[self class] format:@"%@ with error: %@", NSStringFromSelector(_cmd), error];
    self.lastError = error;
    self.returnStatus = kTestRequestStatusDidFail;
}

#pragma mark - SFOAuthCoordinatorDelegate

- (void)oauthCoordinator:(SFOAuthCoordinator *)coordinator willBeginAuthenticationWithView:(WKWebView *)view
{
    NSAssert(NO, @"User Agent flow not supported in this class.");
}

- (void)oauthCoordinator:(SFOAuthCoordinator *)coordinator didStartLoad:(WKWebView *)view
{
    NSAssert(NO, @"User Agent flow not supported in this class.");
}

- (void)oauthCoordinator:(SFOAuthCoordinator *)coordinator didFinishLoad:(WKWebView *)view error:(NSError*)errorOrNil
{
    NSAssert(NO, @"User Agent flow not supported in this class.");
}

- (void)oauthCoordinator:(SFOAuthCoordinator *)coordinator didBeginAuthenticationWithView:(WKWebView *)view
{
    NSAssert(NO, @"User Agent flow not supported in this class.");
}

- (void)oauthCoordinator:(SFOAuthCoordinator *)coordinator didBeginAuthenticationWithSession:(ASWebAuthenticationSession *)session
{
    NSAssert(NO, @"Web Server flow not supported in this class.");
}
- (void)oauthCoordinatorDidCancelBrowserAuthentication:(SFOAuthCoordinator *)coordinator
{
    NSAssert(NO, @"Web Server flow not supported in this class.");
}

- (void)oauthCoordinatorDidBeginNativeAuthentication:(nonnull SFOAuthCoordinator *)coordinator {
    NSAssert(NO, @"Web Server flow not supported in this class.");
}

- (void)oauthCoordinatorDidAuthenticate:(SFOAuthCoordinator *)coordinator authInfo:(SFOAuthInfo *)info
{
    [SFSDKCoreLogger i:[self class] format:@"%@ with authInfo: %@", NSStringFromSelector(_cmd), info];
    self.returnStatus = kTestRequestStatusDidLoad;
}

- (void)oauthCoordinator:(SFOAuthCoordinator *)coordinator didFailWithError:(NSError *)error authInfo:(SFOAuthInfo *)info
{
    [SFSDKCoreLogger i:[self class] format:@"%@ with authInfo: %@, error: %@", NSStringFromSelector(_cmd), info, error];
    self.lastError = error;
    self.returnStatus = kTestRequestStatusDidFail;
}

@end

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Test/SFSDKTestRequestListener.swift
// SFSDKTestRequestListener.swift
// SalesforceSDKCore
//
// Copyright (c) 2012-present, salesforce.com, inc. All rights reserved.
//
// Redistribution and use of this software in source and binary forms, with or without modification,
// are permitted provided that the following conditions are met:
// * Redistributions of source code must retain the above copyright notice, this list of conditions
// and the following disclaimer.
// * Redistributions in binary form must reproduce the above copyright notice, this list of
// conditions and the following disclaimer in the documentation and/or other materials provided
// with the distribution.
// * Neither the name of salesforce.com, inc. nor the names of its contributors may be used to
// endorse or promote products derived from this software without specific prior written
// permission of salesforce.com, inc.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
// IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
// WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import Foundation
import WebKit
import AuthenticationServices

public let kTestRequestStatusWaiting = "waiting"
public let kTestRequestStatusDidLoad = "didLoad"
public let kTestRequestStatusDidFail = "didFail"

@objc public enum SFAccountManagerServiceType: UInt {
    case none = 0
    case oAuth
    case identity
}

@objc(SFSDKTestRequestListener)
@objcMembers
public class SFSDKTestRequestListener: NSObject, SFIdentityCoordinatorDelegate, SFOAuthCoordinatorDelegate {

    @objc public var dataResponse: Any?
    @objc public var lastError: NSError?
    @objc public var returnStatus: String?
    @objc public var maxWaitTime: TimeInterval = 30.0

    private let completionSemaphore = DispatchSemaphore(value: 0)

    public override init() {
        super.init()
        returnStatus = kTestRequestStatusWaiting
    }

    /// Wait for the request to complete (success or fail).
    /// Waits for up to maxWaitTime.
    @objc public func waitForCompletion() -> String {
        // Wait for completion signal with timeout
        let result = completionSemaphore.wait(timeout: .now() + maxWaitTime)
        if result == .timedOut {
            SFSDKCoreLogger.d(SFSDKTestRequestListener.self, message: "Request took too long (> \(maxWaitTime) secs) to complete.")
            return kTestRequestStatusDidFail
        }
        return returnStatus ?? kTestRequestStatusDidFail
    }

    // MARK: - SFIdentityCoordinatorDelegate

    public func identityCoordinatorRetrievedData(_ coordinator: SFIdentityCoordinator) {
        SFSDKCoreLogger.i(SFSDKTestRequestListener.self, message: #function)
        returnStatus = kTestRequestStatusDidLoad
        completionSemaphore.signal()
    }

    public func identityCoordinator(_ coordinator: SFIdentityCoordinator, didFailWithError error: Error) {
        SFSDKCoreLogger.i(SFSDKTestRequestListener.self, message: "\(#function) with error: \(error)")
        lastError = error as NSError
        returnStatus = kTestRequestStatusDidFail
        completionSemaphore.signal()
    }

    // MARK: - SFOAuthCoordinatorDelegate

    public func oauthCoordinator(_ coordinator: SFOAuthCoordinator, willBeginAuthenticationWith view: WKWebView) {
        assertionFailure("User Agent flow not supported in this class.")
    }

    public func oauthCoordinator(_ coordinator: SFOAuthCoordinator, didStartLoad view: WKWebView) {
        assertionFailure("User Agent flow not supported in this class.")
    }

    public func oauthCoordinator(_ coordinator: SFOAuthCoordinator, didFinishLoad view: WKWebView, error: Error?) {
        assertionFailure("User Agent flow not supported in this class.")
    }

    public func oauthCoordinator(_ coordinator: SFOAuthCoordinator, didBeginAuthenticationWithView view: WKWebView) {
        assertionFailure("User Agent flow not supported in this class.")
    }

    public func oauthCoordinator(_ coordinator: SFOAuthCoordinator, didBeginAuthenticationWithSession session: ASWebAuthenticationSession) {
        assertionFailure("Web Server flow not supported in this class.")
    }

    public func oauthCoordinatorDidCancelBrowserAuthentication(_ coordinator: SFOAuthCoordinator) {
        assertionFailure("Web Server flow not supported in this class.")
    }

    public func oauthCoordinatorDidBeginNativeAuthentication(_ coordinator: SFOAuthCoordinator) {
        assertionFailure("Web Server flow not supported in this class.")
    }

    public func oauthCoordinatorDidAuthenticate(_ coordinator: SFOAuthCoordinator, authInfo info: SFOAuthInfo) {
        SFSDKCoreLogger.i(SFSDKTestRequestListener.self, message: "\(#function) with authInfo: \(info)")
        returnStatus = kTestRequestStatusDidLoad
        completionSemaphore.signal()
    }

    public func oauthCoordinator(_ coordinator: SFOAuthCoordinator, didFailWithError error: Error, authInfo info: SFOAuthInfo?) {
        SFSDKCoreLogger.i(SFSDKTestRequestListener.self, message: "\(#function) with authInfo: \(String(describing: info)), error: \(error)")
        lastError = error as NSError
        returnStatus = kTestRequestStatusDidFail
        completionSemaphore.signal()
    }
}
